README.md
# debug-repl

Inspired by [an older
library](https://github.com/georgejahad/debug-repl), debug-repl is a
debug repl implemented as an nrepl middleware. It allows you to set
breakpoints that cause the execution to stop and switches the repl to
evaluate code in the context of the breakpoint.

## Usage

Add the dependency and the middleware, e.g. in your `:user` profile:

``` clojure
:dependencies [[com.gfredericks/debug-repl "0.0.12"]]
:repl-options
  {:nrepl-middleware
    [com.gfredericks.debug-repl/wrap-debug-repl]}
```

Then when you're ready to set breakpoints:

``` clojure
user> (require '[com.gfredericks.debug-repl :refer [break! unbreak!]])
nil
user> (let [x 41] (break!))
Hijacking repl for breakpoint: unnamed
user> (inc x)
42
user> (unbreak!)
nil
nil
```

### `unbreak!!`

The `unbreak!!` function (with two `!`s) will cancel all further
breakpoints for the remainder of the original evaluation's scope.

### `catch-break!`

A macro which will break only if the wrapped code throws an exception.

### async

There's also an async version of the debug-repl, for use in contexts where you cannot call the function directly (e.g. in a web request or worker queue).

``` clojure
user> (require '[com.gfredericks.debug-repl.async :refer [break! unbreak! wait-for-break]])
nil
user> (defn handler [req] (break!) {:status 200})
#'handler
user> (def srv (future (run-jetty handler {:port 8080})))
2019-06-22 13:29:31.400:INFO:oejs.Server:clojure-agent-send-off-pool-0: jetty-9.4.12.v20180830; built: 2018-08-30T13:59:14.071Z; git: 27208684755d94a92186989f695db2d7b21ebc51; jvm 1.8.0_202-b08
#'user/srv
user> (wait-for-breaks) ; At this point I opened my browser to http://localhost:8080
Hijacking repl for breakpoint: unnamed
user> req
{:ssl-client-cert nil, :protocol "HTTP/1.1", :remote-addr "127.0.0.1", :headers {"host" "localhost:8080", "user-agent" "Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0", "cookie" "__stripe_mid=d277c210-219f-4aa7-8fe7-b23715befe83", "connection" "keep-alive", "upgrade-insecure-requests" "1", "accept" "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", "accept-language" "en-GB,en-US;q=0.7,en;q=0.3", "accept-encoding" "gzip, deflate", "dnt" "1", "cache-control" "max-age=0"}, :server-port 8080, :content-length nil, :content-type nil, :character-encoding nil, :uri "/", :server-name "localhost", :query-string nil, :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x6710feb8 "HttpInputOverHTTP@6710feb8[c=0,q=0,[0]=null,s=STREAM]"], :scheme :http, :request-method :get}
user> (unbreak!)
nil
```

## nREPL compatibility

A major change in the nREPL ecosystem around 2018 resulted in a
confusing situation for dev setups involving a composition of
different nREPL-based tools. Because the maven artefact id change
(`org.clojure/nrepl` -> `nrepl/nrepl`) and the base namespace changed
`clojure.tools.nrepl` -> `nrepl`), it is possible to have both the old
version and the new version of nREPL on the classpath at the same
time, with the possibility that some tools expect to use one and some
tools expect to use the other.

debug-repl assumes the old version of nREPL for versions `0.0.9` and
earlier, and supports both as of version `0.0.10`; if it happens that
both versions are on your classpath, debug-repl will optimistically
pick the newer version; however, likely the best situation is to
figure out how to only have the newer version on your classpath.

## TODOs

- Implement `(throw! ex-fn)` for unbreaking and causing the original
  execution to throw an exception.

## License

Copyright © 2014 Gary Fredericks

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


changes.md
# Changelog

## 0.0.12 (2021-01-23)

Thanks to Darrick Wiebe for these features.

- Fix a bug in `catch-break!`.
- Add support for `return!` from `catch-break!`
- Add label support to `catch-break!`

## 0.0.11 (2019-08-04)

Adds the new `com.gfredericks.debug-repl.async` namespace.

## 0.0.10 (2019-02-10)

Supports the namespace change in newer versions of nrepl (and thus
leiningen) ([#7](https://github.com/gfredericks/debug-repl/issues/7)
and [#8](https://github.com/gfredericks/debug-repl/pull/8)).

## 0.0.9 (2017-08-24)

Fixes a [bug with primitive type hints](https://github.com/gfredericks/debug-repl/issues/4).

## 0.0.8

Fixed a race condition in `unbreak!!` that caused it to just not work.

## 0.0.7

Added `catch-break!`.

## 0.0.6

Sorta fix `*1`, `*2`, etc.

## 0.0.5

Detects the NREPL-53 bug instead of failing strangely.

## 0.0.4

Compatibility with clojure 1.4.

## 0.0.3

Add `unbreak!!` for disabling future breaks.


test/com/gfredericks/debug_repl_test.clj
(ns com.gfredericks.debug-repl-test
  (:refer-clojure :exclude [eval])
  (:require [nrepl.core :as client]
            [nrepl.server :as server])
  (:use clojure.test
        com.gfredericks.debug-repl))

(def ^:dynamic *client*)

(defn server-fixture
  [test]
  (with-open [server (server/start-server
                      :port 56408
                      :bind "127.0.0.1"
                      :handler (server/default-handler #'wrap-debug-repl))
              t (client/connect :port 56408 :host "127.0.0.1")]
    (let [c (client/client t 100)]
      (binding [*client* c]
        (test))))
  ;; there's gotta be a better way to wait for the server to really be
  ;; shutdown
  (Thread/sleep 100))

(defn clear-active-repl-info-fixture
  [test]
  (reset! active-debug-repls {})
  (test))

(use-fixtures :each server-fixture clear-active-repl-info-fixture)

(defn fresh-session
  []
  (let [f (client/client-session *client*)]
    (dorun
     (f {:op :eval
         :code (pr-str (list 'ns
                             (gensym "user")
                             '(:require [com.gfredericks.debug-repl :refer [break! unbreak! unbreak!! catch-break! return!]])))}))
    f))

(defn unparsed-eval*
  [session-fn code-string]
  (->> (session-fn {:op :eval, :code code-string})
       (map (fn [{:keys [ex err] :as msg}]
              (if (or ex err)
                (throw (ex-info (str "Error during eval!" (pr-str msg))
                                {:msg msg}))
                msg)))
       (keep :value)
       (doall)))

(defn eval*
  [session-fn code-string]
  (map read-string (unparsed-eval* session-fn code-string)))

(defmacro eval
  "Returns a sequence of return values from the evaluation."
  [session-fn eval-code]
  `(eval* ~session-fn (client/code ~eval-code)))

(defmacro eval-raw
  [session-fn eval-code]
  `(~session-fn {:op :eval, :code (client/code ~eval-code)}))

(deftest hello-world-test
  (let [f (fresh-session)]
    (is (= [] (eval f (let [x 42] (break!) :return)))
        "Breaking returns no results.")
    (is (= [42] (eval f x))
        "Evaluation sees the context of the break.")
    ;; relaxing the ordering for now until I have a coherent design
    ;; idea about it.
    (is (= #{:return nil} (set (eval f (unbreak!))))
        "unbreak first returns the return value from the
         unbroken thread, then its own nil.")))

(deftest break-out-of-loop-test
  (let [f (fresh-session)]
    (is (= [] (eval f (do (dotimes [n 10] (break!)) :final-return))))
    (is (= [0] (eval f n)))
    (is (= [nil] (eval f (unbreak!))))
    (is (= [1] (eval f n)))
    (is (= #{:final-return nil} (set (eval f (unbreak!!)))))
    ;; should be able to break again now
    (is (= [] (eval f (let [x 42] (break!) :return))))
    (is (= [42] (eval f x)))
    (is (= #{:return nil} (set (eval f (unbreak!)))))))

(deftest repl-vars-test
  (let [f (fresh-session)]
    (testing "*1"
      (is (= [] (eval f (let [y 7] (break!) :return))))
      (is (= [42] (eval f (* 2 3 y))))
      (is (= [42] (eval f *1)))

      ;; it would be nice if this worked but also seems hard
      ;; to do cleanly
      #_#_
      (is (= #{nil :return} (set (eval f (unbreak!)))))
      (let [[xs] (eval f [*1 *2 *3])]
        (is (some #{42} xs))))

    (testing "*e"
      (let [[msg1 msg2] (eval-raw f (/ 42 0))]
        (is (= (clojure.set/subset? #{:err :ex} (set (concat (keys msg1) (keys msg2)))))))
      (is (= ["java.lang.ArithmeticException"]
             (eval f (-> *e class .getName)))))

    (testing "Unbound primitives"
      (is (unparsed-eval* f "(fn [^long x] (break!))")))))

(deftest catch-break-regression-test
  (let [f (fresh-session)]
    (eval f (catch-break! (throw (Exception. "oh well"))))
    (is (thrown-with-msg? Exception #"oh well" (eval f (unbreak!))))))

(deftest return!-test
  (let [f (fresh-session)]
    (eval f (def jake (atom nil)))
    (eval f (reset! jake (catch-break! (throw (Exception. "oh well")))))
    (eval f (return! :twelve))
    (is (= [:twelve] (eval f @jake)))))


project.clj
(defproject com.gfredericks/debug-repl "0.0.13-SNAPSHOT"
  :description "A Clojure debug repl as nrepl middleware."
  :url "https://github.com/fredericksgary/debug-repl"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]]
  :deploy-repositories [["releases" :clojars]]
  :profiles {:1.3 {:dependencies [^:replace [org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [^:replace [org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [^:replace [org.clojure/clojure "1.5.1"]]}
             :1.6 {:dependencies [^:replace [org.clojure/clojure "1.6.0"]]}
             :1.7 {:dependencies [^:replace [org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [^:replace [org.clojure/clojure "1.8.0"]]}
             :old-lein {:plugins [[com.gfredericks/nrepl-53-monkeypatch "0.1.0"]]}}
  ;; use `lein all test` to run the tests on all versions
  ;;
  ;; Skipping 1.3 because I'm not in the mood to do backpat tricks for
  ;; ex-info.
  :aliases {"all" ["with-profile" "+1.4:+1.5:+1.6:+1.7:+1.8"]})


src/debug_repl/core.clj
(ns debug-repl.core
  (:require [clojure.tools.nrepl.middleware :refer [set-descriptor!]]
            [clojure.tools.nrepl.middleware.interruptible-eval :refer [*msg*]]
            [clojure.tools.nrepl.misc :refer [response-for]]
            [clojure.tools.nrepl.transport :as transport]))

;; TODO:
;;   - Report the correct ns so the repl switches back & forth?
;;   - Avoid reporting :done multiple times

(defonce
  ^{:doc
    "A map from nrepl session IDs to a stack of debug repl maps, each of which
     contain:

    :unbreak -- a promise which will cause the thread of execution to resume
                when it is delivered
    :nested-session-id -- the nrepl session ID being used to evaluate code
                          for this repl
    :eval -- a function that takes a code string and returns the result of
             evaling in this repl."}
  active-debug-repls
  (atom {}))

(defmacro current-locals
  "Returns a map from symbols of locals in the lexical scope to their
  values."
  []
  (into {}
        (for [name (keys &env)]
          [(list 'quote name) name])))

(defmacro ^:private catchingly
  "Returns either [:returned x] or [:threw t]."
  [& body]
  `(try [:returned (do ~@body)]
        (catch Throwable t#
          [:threw t#])))

(defn ^:private uncatch
  [[type x]]
  (case type :returned x :threw (throw x)))

(defn break
  [locals breakpoint-name ns]
  (let [{:keys [transport],
         session-id ::orig-session-id
         nest-session-fn ::nest-session}
        *msg*

        unbreak-p (promise)
        eval-requests (atom clojure.lang.PersistentQueue/EMPTY)
        pop-eval-request (fn []
                           (when-not (empty? @eval-requests)
                             (-> eval-requests
                                 (swap! (fn [q]
                                          (if (empty? q)
                                            q
                                            (vary-meta (pop q) assoc :popped (peek q)))))
                                 (meta)
                                 :popped)))]
    ;; TODO: nesting
    (swap! active-debug-repls update-in [session-id] conj
           {:unbreak           unbreak-p
            :nested-session-id (nest-session-fn)
            :eval              (fn [code]
                                 (let [result-p (promise)]
                                   (swap! eval-requests conj [code result-p])
                                   (uncatch @result-p)))})
    (transport/send transport
                    (response-for *msg*
                                  {:out (str "Hijacking repl for breakpoint: "
                                             breakpoint-name)}))
    (transport/send transport
                    (response-for *msg*
                                  {:status #{:done}}))
    (loop []
      (when-not (realized? unbreak-p)
        (if-let [[code result-p] (pop-eval-request)]
          (let [code' (format "(fn [{:syms [%s]}]\n%s\n)"
                              (clojure.string/join " " (keys locals))
                              code)]
            (deliver result-p
                     (catchingly
                      ((binding [*ns* ns] (eval (read-string code'))) locals))))
          (Thread/sleep 50))
        (recur)))
    nil))

(defmacro break!
  ([]
     `(break! "unnamed"))
  ([breakpoint-name]
     `(break (current-locals)
             ~breakpoint-name
             ~*ns*)))

(defn unbreak!
  []
  (let [{session-id ::orig-session-id} *msg*
        p (-> @active-debug-repls
              (get session-id)
              (peek)
              (:unbreak))]
    (when-not p
      (throw (Exception. "No debug-repl to unbreak from!")))
    ;; TODO: dissoc as well? (minor memory leak)
    (swap! active-debug-repls update-in [session-id] pop)
    (deliver p nil)
    nil))

(defn wrap-transport-sub-session
  [t from-session to-session]
  (reify transport/Transport
    (recv [this] (transport/recv t))
    (recv [this timeout] (transport/recv t timeout))
    (send [this msg]
      (let [msg' (cond-> msg (= from-session (:session msg)) (assoc :session to-session))]
        (transport/send t msg')))))

(defn ^:private wrap-eval
  [{:keys [op code session] :as msg}]
  (let [nested-session-id (-> @active-debug-repls
                              (get session)
                              (peek)
                              (:nested-session-id))]
    (cond-> msg
            nested-session-id
            (-> (assoc :session nested-session-id)
                (update-in [:transport] wrap-transport-sub-session nested-session-id session))


            (and nested-session-id (= "eval" op))
            (assoc :code
              (pr-str
               `((-> @active-debug-repls
                     (get ~session)
                     (peek)
                     (:eval))
                 ~code))))))

(defn handle-debug
  [handler {:keys [transport op code session] :as msg}]
  (-> msg
      (assoc ::orig-session-id session
             ::nest-session (fn []
                              {:post [%]}
                              (let [p (promise)]
                                (handler {:session session,
                                          :op "clone",
                                          :transport (reify transport/Transport
                                                       (send [_ msg]
                                                         (deliver p msg)))})
                                (:new-session @p))))
      (wrap-eval)
      (handler)))

(defn wrap-debug-repl
  [handler]
  ;; having handle-debug as a separate function makes it easier to do
  ;; interactive development on this middleware
  (fn [msg] (handle-debug handler msg)))

(defn try-it
  [x]
  (break!))


src/com/gfredericks/debug_repl/async.clj
(ns com.gfredericks.debug-repl.async
  (:require
    [com.gfredericks.debug-repl :as debug-repl]
    [com.gfredericks.debug-repl.util :as util])
  (:import (java.util.concurrent
            SynchronousQueue
            TimeUnit)))

(def the-executor nil) ;; global mutable state

(def ^:private msg-var
  (if (util/require? 'nrepl.server)
    (resolve 'nrepl.middleware.interruptible-eval/*msg*)
    (resolve 'clojure.tools.nrepl.middleware.interruptible-eval/*msg*)))

(defn can-break?
  "Implementation detail. Subject to change."
  []
  ;; this just checks if we're in a repl
  (boolean (var-get msg-var)))

(defmacro break!
  "Equivalent to com.gfredericks.debug-repl/break! except that if not run from
  an nREPL session (e.g. from a ring request) then will attempt to connect to
  the connection where wait-for-breaks is being called."
  [& args]
  `(if (can-break?)
     (debug-repl/break! ~@args)
     (if-let [e# the-executor]
       (e# #(debug-repl/break! ~@args))
       :noop)))

(defmacro catch-break!
  "Executes body and breaks if it throws an exception. The exception
  will be in the local scope as &ex. The exception will be re-thrown
  after unbreaking."
  [& body]
  (let [[name body]
        (if (and (or (string? (first body)) (keyword? (first body)))
                 (next body))
          [(first body) (next body)]
          ["catch-break!" body])]
    `(try ~@body (catch Throwable ~'&ex
                   (swap! debug-repl/break-return conj ::throw)
                   (break! ~name)
                   (let [return# (peek @debug-repl/break-return)]
                     (swap! break-return pop)
                     (if (= ::throw return#)
                       (throw ~'&ex)
                       return#))))))

(defn wait-for-breaks
  "Wait for a call to break! outside of the nREPL.  Takes an optional timeout
  in seconds to wait which is by default 10 seconds.
  
  The second arg is a boolean (defaulting to true) that determines whether 
  calls to `break!` should block when other break calls are currently active."
  ([] (wait-for-breaks 10 true))
  ([timeout] (wait-for-breaks timeout true))
  ([timeout-seconds wait?]
   (let [queue (SynchronousQueue.)

         executor
         (fn [func]
           (let [p (promise)
                 f (bound-fn
                     []
                     (deliver p
                              (try
                                [(func)]
                                (catch Throwable t
                                  [nil t]))))]

             (if (or (and wait? (do (.put queue f) true))
                     (.offer queue f))
               (let [[x err] @p]
                 (if err (throw err) x))
               :noop)))]

     (with-redefs [the-executor (or the-executor executor)]
       (when (= the-executor executor)
         (loop []
           (if-let [func (.poll queue timeout-seconds TimeUnit/SECONDS)]
             (do
               (func)
               (recur))
             :no-reqs-before-timeout)))))))


src/com/gfredericks/debug_repl/http_intercept.clj
(ns com.gfredericks.debug-repl.http-intercept
  (:require
    [com.gfredericks.debug-repl :as debug-repl]
    [com.gfredericks.debug-repl.util :as util])
  (:import (java.util.concurrent
            ArrayBlockingQueue
            TimeUnit)))

(def the-executor nil) ;; global mutable state

(defn can-break?
  []
  ;; this just checks if we're in a repl
  (boolean
    (if (util/require? 'nrepl.server)
      @(resolve 'nrepl.middleware.interruptible-eval/*msg*)
      @(resolve 'clojure.tools.nrepl.middleware.interruptible-eval/*msg*))))

(defmacro break!
  [& args]
  `(if (can-break?)
     (debug-repl/break! ~@args)
     (if-let [e# the-executor]
       (e# #(debug-repl/break! ~@args))
       :noop)))

(defn wait-for-breaks
  []
  ;; if the repl is busy executing a request, should concurrent
  ;; requests block or skip the repl execution altogether?
  ;;
  ;; going to try blocking first
  (let [queue (java.util.concurrent.ArrayBlockingQueue. 16)

        executor
        (fn [func]
          (let [p (promise)
                f (bound-fn
                    []
                    (deliver p
                             (try
                               [(func)]
                               (catch Throwable t
                                 [nil t]))))]
            (.put queue f)
            (let [[x err] @p]
              (if err (throw err) x))))]

    (alter-var-root #'the-executor #(or % executor))
    (when (= the-executor executor)
      (try
        (loop []
          (if-let [func (.poll queue 10 TimeUnit/SECONDS)]
            (do
              (func)
              (recur))
            :no-reqs-for-10-seconds))
        (finally
          (alter-var-root #'the-executor (constantly nil)))))))


src/com/gfredericks/debug_repl.clj
(ns com.gfredericks.debug-repl
  ;; backwards compatibility stuff
  (:refer-clojure :exclude [cond->])
  (:require [com.gfredericks.debug-repl.backpat :refer [cond->]])

  ;; normal requires
  (:require [com.gfredericks.debug-repl.nrepl-53 :refer [report-nrepl-53-bug]]
            [com.gfredericks.debug-repl.util :as util])
  (:import (java.util.concurrent ArrayBlockingQueue)))

(if (util/require? 'nrepl.server)
  (require '[nrepl.middleware :refer [set-descriptor!]]
           '[nrepl.middleware.interruptible-eval :refer [*msg*]]
           '[nrepl.misc :refer [response-for]]
           '[nrepl.transport :as transport])
  (require '[clojure.tools.nrepl.middleware :refer [set-descriptor!]]
           '[clojure.tools.nrepl.middleware.interruptible-eval :refer [*msg*]]
           '[clojure.tools.nrepl.misc :refer [response-for]]
           '[clojure.tools.nrepl.transport :as transport]))

;; TODO:
;;   - Close nrepl sessions after unbreak!
;;   - Report the correct ns so the repl switches back & forth?
;;   - Avoid reporting :done multiple times
;;   - Suppress the return value from (unbreak!)? this would avoid
;;     the command returning two results...
;;   - Detect when (break!) is called but the middleware is missing?
;;     And give a helpful error message.
;;   - Better reporting on how many nested repls there are, etc

(defonce
  ^{:doc
    "A map from nrepl session IDs to a stack of debug repl maps, each of which
     contain:

    :unbreak -- a 0-arg function which will cause the thread of
                execution to resume when it is called
    :nested-session-id -- the nrepl session ID being used to evaluate code
                          for this repl
    :eval -- a function that takes a code string and returns the result of
             evaling in this repl."}
  active-debug-repls
  (atom {}))

(defn ^:private set-no-more-breaks!
  "Sets the flag in this debug repl so that it will not break anymore
  until the original eval is complete."
  [session-id msg-id]
  (swap! active-debug-repls assoc-in [session-id :no-more-breaks?] msg-id))

(defn ^:private maybe-clear-no-more-breaks!
  "Clears no-more-breaks unless it was set by this message."
  [session-id msg-id]
  (swap! active-debug-repls update-in [session-id]
         (fn [session-data]
           (cond-> session-data
             (not= (get session-data :no-more-breaks?) msg-id)
             (dissoc :no-more-breaks?)))))

(defn ^:private no-more-breaks?
  [session-id]
  (get-in @active-debug-repls [session-id :no-more-breaks?]))

(defn innermost-debug-repl
  [session-id]
  (peek (get-in @active-debug-repls [session-id :repls])))

(defmacro current-locals
  "Returns a map from symbols of locals in the lexical scope to their
  values."
  []
  (into {}
        (for [name (keys &env)]
          [(list 'quote name)
           (vary-meta name dissoc :tag) ])))


(defn break
  [locals breakpoint-name ns]
  (let [{:keys [transport],
         session-id ::orig-session-id
         nest-session-fn ::nest-session}
        *msg*

        unbreak-p (promise)
        ;; probably never need more than 1 here
        eval-requests (ArrayBlockingQueue. 2)]
    (when-not (no-more-breaks? session-id)
      (swap! active-debug-repls update-in [session-id :repls] conj
             {:unbreak           (fn [] (deliver unbreak-p nil))
              :nested-session-id (nest-session-fn)
              :eval              (fn [code]
                                   (let [result-p (promise)
                                         ;; using the bindings from
                                         ;; the cloned session seems
                                         ;; like the best way to get
                                         ;; *1, *2, etc.  to work
                                         ;; right. Not sure if there
                                         ;; are other surprising
                                         ;; consequences.
                                         binding-fn (bound-fn [f] (f))]
                                     (.put eval-requests [code binding-fn result-p])
                                     (util/uncatch @result-p)))})
      (transport/send transport
                      (response-for *msg*
                                    {:out (str "Hijacking repl for breakpoint: "
                                               breakpoint-name)}))
      (transport/send transport
                      (response-for *msg*
                                    {:status #{:done}}))
      (loop []
        (when-not (realized? unbreak-p)
          (if-let [[code binding-fn result-p] (.poll eval-requests)]
            (let [code' (format "(fn [{:syms [%s]}]\n%s\n)"
                                (clojure.string/join " " (keys locals))
                                code)]
              (deliver result-p
                       (util/catchingly
                        (binding-fn
                         (fn []
                           ((binding [*ns* ns] (eval (read-string code'))) locals))))))
            (Thread/sleep 50))
          (recur))))
    nil))

(defmacro break!
  "Use only with the com.gfredericks.debug-repl/wrap-debug-repl middleware.

  Causes execution to stop and the repl switches to evaluating code in the
  context of the breakpoint. Resume execution by calling (unbreak!). REPL
  code can result in a nested call to break! which will work in a reasonable
  way. Nested breaks require multiple calls to (unbreak!) to undo."
  ([]
     `(break! "unnamed"))
  ([breakpoint-name]
     `(break (current-locals)
             ~breakpoint-name
             ~*ns*)))

(defn unbreak!
  "Causes the latest breakpoint to resume execution; the repl returns to the
  state it was in prior to the breakpoint."
  []
  (let [{session-id ::orig-session-id} *msg*
        f (:unbreak (innermost-debug-repl session-id))]
    (when-not f
      (throw (Exception. "No debug-repl to unbreak from!")))
    ;; TODO: dissoc as well? (minor memory leak)
    (swap! active-debug-repls update-in [session-id :repls] pop)
    (f)
    nil))

(defn unbreak!!
  "Like unbreak! but cancels all remaining breakpoints for the
  original evaluation."
  []
  (set-no-more-breaks! (::orig-session-id *msg*) (::msg-id *msg*))
  (unbreak!))

(defn ^:private wrap-transport-sub-session
  [t from-session to-session]
  (reify transport/Transport
    (recv [this] (transport/recv t))
    (recv [this timeout] (transport/recv t timeout))
    (send [this msg]
      (let [msg' (cond-> msg (= from-session (:session msg)) (assoc :session to-session))]
        (transport/send t msg')))))

(defn ^:private wrap-eval
  [{:keys [op code session] :as msg}]
  (let [{:keys [nested-session-id]} (innermost-debug-repl session)]
    (cond-> msg
            nested-session-id
            (-> (assoc :session nested-session-id)
                (update-in [:transport] wrap-transport-sub-session nested-session-id session))


            (and nested-session-id (= "eval" op))
            (assoc :code
              (pr-str
               `((:eval (innermost-debug-repl ~session))
                 ~code))))))

(defn ^:private wrap-transport-cleanup
  [t session-id msg-id]
  (reify transport/Transport
    (recv [this] (transport/recv t))
    (recv [this timeout] (transport/recv t timeout))
    (send [this msg]
      (when (and (:done (:status msg))
                 (let [m (get @active-debug-repls session-id)]
                   (and m (empty? (:repls m)))))
        (maybe-clear-no-more-breaks! session-id msg-id))
      (transport/send t msg))))

(defn syncronous-new-session
  [handler session-id]
  (let [p (promise)]
    (handler {:session session-id
              :op "clone"
              :transport (reify transport/Transport
                           (send [_ msg]
                             (deliver p msg)))})
    (when (:unknown-op (:status @p))
      (throw (ex-info "Bad middleware ordering!" {:type ::bad-middleware-ordering})))
    (:new-session @p)))

(defn ^:private handle-debug
  [handler {:keys [transport op code session] :as msg}]
  (let [msg-id (java.util.UUID/randomUUID)]
    (-> msg
        (assoc ::orig-session-id session
               ::msg-id msg-id
               ::nest-session (fn []
                                {:post [%]}
                                (syncronous-new-session handler session)))
        (update-in [:transport] wrap-transport-cleanup session msg-id)
        (wrap-eval)
        (handler))))

(defn wrap-debug-repl
  [handler]
  ;; Test for NREPL-53 at startup
  (try (syncronous-new-session handler nil)
       (catch clojure.lang.ExceptionInfo e
         (when (= ::bad-middleware-ordering (:type (ex-data e)))
           (report-nrepl-53-bug))))

  ;; having handle-debug as a separate function makes it easier to do
  ;; interactive development on this middleware
  (fn [msg] (handle-debug handler msg)))

(set-descriptor! #'wrap-debug-repl
                 {:expects #{"eval" "clone"}})

;;
;; Helpers
;;

;; this being a global atom is probably a problem for particular
;; complicated uses; it could probably be fixed by making the var
;; dynamic and creating a new atom with each catch-break call; the
;; only detail to pay attention to is the binding conveyance, making
;; sure that the atom is visible on whatever thread actually executes
;; the return! call
(def break-return (atom []))

(defn return! [value]
  (swap! break-return #(-> % pop (conj value)))
  (unbreak!))

(defn return!! [value]
  (swap! break-return #(-> % pop (conj value)))
  (unbreak!!))

(defmacro catch-break!
  "Executes body and breaks if it throws an exception. The exception
  will be in the local scope as &ex. The exception will be re-thrown
  after unbreaking."
  [& body]
  (let [[name body]
        (if (and (or (string? (first body)) (keyword? (first body)))
                 (next body))
          [(first body) (next body)]
          ["catch-break!" body])]
    `(try ~@body (catch Throwable ~'&ex
                   (swap! break-return conj ::throw)
                   (break! ~name)
                   (let [return# (peek @break-return)]
                     (swap! break-return pop)
                     (if (= ::throw return#)
                       (throw ~'&ex)
                       return#))))))
